lib/Builtins.cpp

```diff
@@ -14,171 +14,335 @@
+  case 'k': // ??? not part of cxxabi
+  case 'K': // const
+  case 'V': // volatile
+    return GetParameterType(mangled_name, type_info, pos);
+  case 'U': { // Address space
+    std::string address_space = GetUnmangledName(mangled_name, &pos);
+    return GetParameterType(mangled_name, type_info, pos);
+  }
+    // OCL types
+  case 'D':
+    type_code = mangled_name[pos++];
+    if (type_code == 'v') { // OCL vector
+      char *end;
+      int numElems = strtol(&mangled_name[pos], &end, 10);
+      if (!numElems) {
+        return 0;
+      }
+      type_info->vector_size = numElems;
+      pos = end - mangled_name.data();
+
+      if (mangled_name[pos++] != '_') {
+        return 0;
+      }
+      return GetParameterType(mangled_name, type_info, pos);
+    } else if (type_code == 'h') { // OCL half
+      type_info->type_id = Type::FloatTyID;
+      type_info->is_signed = true;
+      type_info->byte_len = 2;
+    } else {
+#ifdef DEBUG
+      printf("Func: %s\n", mangled_name.c_str());
+      llvm_unreachable("failed to demangle name");
+#endif
+      return 0;
+    }
+    break;
+
+    // element types
+  case 'l':
+    blen <<= 1; // long
+  case 'i':
+    blen <<= 1; // int
+  case 's':
+    blen <<= 1; // short
+  case 'c':     // char
+  case 'a':     // signed char
+    type_info->type_id = Type::IntegerTyID;
+    type_info->is_signed = true;
+    type_info->byte_len = blen;
+    break;
+  case 'm':
+    blen <<= 1; // unsigned long
+  case 'j':
+    blen <<= 1; // unsigned int
+  case 't':
+    blen <<= 1; // unsigned short
+  case 'h':     // unsigned char
+    type_info->type_id = Type::IntegerTyID;
+    type_info->is_signed = false;
+    type_info->byte_len = blen;
+    break;
+  case 'd':
+    blen = 2; // double float
+  case 'f':
+    blen <<= 2; // single float
+    type_info->type_id = Type::FloatTyID;
+    type_info->is_signed = true;
+    type_info->byte_len = blen;
+    break;
+  case 'v': // void
+    break;
+  case '1':
+  case '2':
+  case '3':
+  case '4':
+  case '5':
+  case '6':
+  case '7':
+  case '8':
+  case '9': {
+    type_info->type_id = Type::StructTyID;
+    pos--;
+    type_info->name = GetUnmangledName(mangled_name, &pos);
+    break;
+  }
+  case '.':
+    return 0;
+  default:
+#ifdef DEBUG
+    printf("Func: %s\n", mangled_name.c_str());
+    llvm_unreachable("failed to demangle name");
+#endif
+    return 0;
+  }
+  return pos;
+}
+} // namespace
+
+////////////////////////////////////////////////////////////////////////////////
+// FunctionInfo::GetFromMangledNameCheck
+//   - parse mangled name as far as possible. Some names are an aggregate of
+//   fields separated by '.'
+//   - extract name and parameter types, and return type for 'convert' functions
+//   - return true if the mangled name can be fully parsed
+bool Builtins::FunctionInfo::GetFromMangledNameCheck(
+    const std::string &mangled_name) {
+  size_t pos = 0;
+  if (mangled_name[pos++] != '_' || mangled_name[pos++] != 'Z') {
```

Comment:
Buffer overrun:  Protect against an input like "X", where you end up indexing too far.

Reply:
Should short circuit if mangled_name[0] != '_' but I can reverse the logic to make it more clear that the first 2 characters must equal "_Z".